### DISCUSSION TITLE
how to get access from another service

### DISCUSSION THREAD

**Discussion by danilkorotkov** [HomeSpan/HomeSpan#61] [Q&A] [answered]
how to get pointer to Service?
For example 
new SL_GATE(); where struct SL_GATE : Service::GarageDoorOpener

SL_GATE has definition SpanService *GatePosition; 

and then GatePosition = new GateDoor(); where struct GateDoor : Service::Door
no error when compiling

but if i add smth ...GatePosition->CurrentPosition->getVal() - drops error  No CurrentPosition

of course CurrentPosition works properly inside of GateDoor
```
GateDoor() : Service::Door(){
      CurrentPosition      = new Characteristic::CurrentPosition(50);
      TargetPosition       = new Characteristic::TargetPosition(50);
      PositionState        = new Characteristic::PositionState(2);
      ObstructionDetected  = new Characteristic::ObstructionDetected(false);
      HoldPosition         = new Characteristic::HoldPosition(false);
```

**Comment by HomeSpan** [comment, marked as answer]
In C++, if you declare a variable to be a pointer to the parent structure, it does not have access to any members variables of derived child-classes, even though the child-class structures can (because of inheritance) access parent-class variables (unless they are declared as private).

In your example above, if GatePosition is defined as a pointer to SpanService, it therefore only has access to variables declared in SpanService, not any members declared in structures you derived from SpanService.  The solution is to define GatePosition as a pointer to your *child class* instead:

```
GateDoor *GatePosition = new GateDoor();
```

This way the compiler knows that GatePosition has all of the members of GateDoor (as well as access to all of its parent classes through inheritance).
